src/Microsoft.VisualStudio.Editor.Razor/BraceSmartIndenter.cs

```diff
@@ -0,0 +1,218 @@
+            {
+                throw new ArgumentNullException(nameof(textBuffer));
+            }
+
+            if (editorOperationsFactory == null)
+            {
+                throw new ArgumentNullException(nameof(editorOperationsFactory));
+            }
+
+            _dispatcher = dispatcher;
+            _textBuffer = textBuffer;
+            _editorOperationsFactory = editorOperationsFactory;
+            _textBuffer.Changed += TextBuffer_OnChanged;
+            _textBuffer.PostChanged += TextBuffer_OnPostChanged;
+        }
+
+        public void Dispose()
+        {
+            _dispatcher.AssertForegroundThread();
+
+            _textBuffer.Changed -= TextBuffer_OnChanged;
+            _textBuffer.PostChanged -= TextBuffer_OnPostChanged;
+        }
+
+        private void TextBuffer_OnPostChanged(object sender, EventArgs e)
+        {
+            _dispatcher.AssertForegroundThread();
+
+            var context = _context;
+            _context = null;
+
+            if (context != null)
+            {
+                // Save the caret position
+                var textView = context.FocusedTextView;
+                var caret = textView.Caret.Position.BufferPosition;
+
+                // Insert the completion text, i.e. "\r\n      "
+                var edit = textView.TextBuffer.CreateEdit();
+                var indent = CalculateIndent(textView.TextBuffer, context.ChangePosition);
+                edit.Insert(caret.Position, indent);
+                edit.Apply();
+
+                // Place the caret inbetween the braces (before our indent).
+                var currentSnapshotPoint = new SnapshotPoint(textView.TextBuffer.CurrentSnapshot, caret.Position);
+                textView.Caret.MoveTo(currentSnapshotPoint);
+
+                // For Razor metacode cases the editors smart indent wont kick in automatically. This forces the smart indent.
+                // For example attempting to do enter a newline between the functions directive: @functions {} will not auto-indent
+                // inbetween the braces unless we forcefully move to end of line.
+                var editorOperations = _editorOperationsFactory.GetEditorOperations(textView);
+                editorOperations.MoveToEndOfLine(false);
+            }
+        }
+
+        // Internal for testing
+        internal void TextBuffer_OnChanged(object sender, TextContentChangedEventArgs contentChange)
+        {
+            _dispatcher.AssertForegroundThread();
+
+            if (contentChange.Changes.Count == 0)
+            {
+                return;
+            }
+
+            var firstChange = contentChange.Changes[0];
+            var lastChange = contentChange.Changes[contentChange.Changes.Count - 1];
+            var oldLength = lastChange.OldEnd - firstChange.OldPosition;
+            var newLength = lastChange.NewEnd - firstChange.NewPosition;
+            var newText = contentChange.After.GetText(firstChange.NewPosition, newLength);
```

Comment:
Can we put this in a private method called `GetNormalizedChange` - if this is duplicated with the parser maybe an extension method?